internal/services/tesla_service.go

```diff
@@ -82,3 +86,101 @@ type TeslaVehicle struct {
 	VehicleID int    `json:"vehicle_id"`
 	VIN       string `json:"vin"`
 }
+
+type GetVehiclesResponse struct {
+	Response *[]TeslaVehicle `json:"response"`
+}
+
+type TeslaError struct {
+	Error            string
+	ErrorDescription string `json:"error_description"`
+	ReferenceID      string
+}
+
+var teslaScopes = []string{"user_data", "vehicle_device_data", "vehicle_cmds", "vehicle_charging_cmds", "energy_device_data", "energy_device_data", "energy_cmds"}
+
+type TeslaAuthCodeResponse struct {
+	AccessToken  string `json:"access_token"`
+	RefreshToken string `json:"refresh_token"`
+	IDToken      string `json:"id_token"`
+	ExpiresIn    int    `json:"expires_in"`
+	State        string `json:"state"`
+	TokenType    string `json:"token_type"`
+}
+
+func (t *teslaService) CompleteTeslaAuthCodeExchange(authCode, redirectURI, region string) (*TeslaAuthCodeResponse, error) {
+	u, _ := url.Parse("https://auth.tesla.com/oauth2/v3/token")
+	f := url.Values{}
+	f.Set("grant_type", "authorization_code")
+	f.Set("client_id", t.Settings.TeslaAuthorization.ClientID)
+	f.Set("client_secret", t.Settings.TeslaAuthorization.ClientSecret)
+	f.Set("code", authCode)
+	f.Set("redirect_uri", redirectURI)
+	f.Set("scope", "openid offline_access "+strings.Join(teslaScopes, " "))
+	f.Set("audience", fmt.Sprintf("https://fleet-api.prd.%s.vn.cloud.tesla.com", region))
+
+	req, err := http.NewRequest("POST", u.String(), strings.NewReader(f.Encode()))
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Add("Content-Type", "application/x-www-form-urlencoded")
+
+	res, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer res.Body.Close()
+
+	if res.StatusCode != http.StatusOK {
+		errBody := new(TeslaError)
+		if err := json.NewDecoder(res.Body).Decode(errBody); err != nil {
+			return nil, err
+		}
+		return nil, fmt.Errorf("error occurred completing authorization: %s", errBody.ErrorDescription)
+	}
+
+	teslaAuth := new(TeslaAuthCodeResponse)
+	if err := json.NewDecoder(res.Body).Decode(teslaAuth); err != nil {
+		return nil, err
+	}
+
+	return teslaAuth, nil
+}
+
+func (t *teslaService) GetVehicles(token, region string) ([]TeslaVehicle, error) {
+	u := &url.URL{
+		Scheme: "https",
+		Host:   fmt.Sprintf("fleet-api.prd.%s.vn.cloud.tesla.com", region),
+		Path:   "api/1/vehicles",
+	}
+	req, err := http.NewRequest("GET", u.String(), nil)
```

Comment:
Same as above for using a context
```suggestion
	ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
	defer cancel()
	req, err : = http.NewRequestWithContext(ctx, http.MethodGet, u.String(), nil)
```